File: apps/extension/src/shared/job-board-detector.ts
import type { BoardFeatures } from "./types";

// ── URL Patterns ─────────────────────────────────────────────────────────────

const JOB_BOARD_PATTERNS: Record<string, RegExp[]> = {
  greenhouse: [/greenhouse\.io/i, /boards\.greenhouse\.io/i, /job-boards\.greenhouse\.io/i],
  workday: [/workday\.com/i, /myworkdayjobs\.com/i],
  lever: [/lever\.co/i, /jobs\.lever\.co/i],
  smartrecruiters: [/smartrecruiters\.com/i],
  icims: [/icims\.com/i],
  taleo: [/taleo\.net/i, /taleo\.com/i],
  jobvite: [/jobvite\.com/i],
  bamboohr: [/bamboohr\.com/i],
  linkedin: [/linkedin\.com\/jobs/i],
  indeed: [/indeed\.com/i],
  glassdoor: [/glassdoor\.com/i],
};

// ── Feature Matrix ───────────────────────────────────────────────────────────

const BOARD_FEATURES: Record<string, BoardFeatures> = {
  greenhouse: {
    auto_fill_supported: true,
    file_upload_supported: true,
    custom_questions: true,
    requires_manual_review: false,
  },
  workday: {
    auto_fill_supported: true,
    file_upload_supported: true,
    custom_questions: true,
    requires_manual_review: true,
  },
  lever: {
    auto_fill_supported: true,
    file_upload_supported: true,
    custom_questions: true,
    requires_manual_review: false,
  },
  smartrecruiters: {
    auto_fill_supported: true,
    file_upload_supported: true,
    custom_questions: true,
    requires_manual_review: false,
  },
  linkedin: {
    auto_fill_supported: true,
    file_upload_supported: true,
    custom_questions: false,
    requires_manual_review: false,
  },
  generic: {
    auto_fill_supported: true,
    file_upload_supported: true,
    custom_questions: true,
    requires_manual_review: true,
  },
};

// ── Public API ───────────────────────────────────────────────────────────────

/** Detect job board type from a URL string. Returns 'generic' if no match. */
export function detectJobBoard(url: string): string {
  for (const [board, patterns] of Object.entries(JOB_BOARD_PATTERNS)) {
    if (patterns.some((re) => re.test(url))) {
      return board;
    }
  }
  return "generic";
}

/** Get board type and feature matrix for a URL. */
export function getJobBoardInfo(url: string): { type: string; features: BoardFeatures } {
  const type = detectJobBoard(url);
  const features = BOARD_FEATURES[type] ?? BOARD_FEATURES.generic;
  return { type, features };
}

/** Check if a URL belongs to a known ATS (non-generic). */
export function isKnownJobBoard(url: string): boolean {
  return detectJobBoard(url) !== "generic";
}

File: apps/api/src/core/auto_application/job_board_detector.py
"""Job board detector — extracted from src/auto_application/job_board_detector.py.

Pure regex-based detection. No external dependencies.
"""

import re
from typing import Dict

JOB_BOARD_PATTERNS: Dict[str, list] = {
    "greenhouse": [r"greenhouse\.io", r"boards\.greenhouse\.io", r"job-boards\.greenhouse\.io"],
    "workday": [r"workday\.com", r"myworkdayjobs\.com", r"wd\d+\.myworkdayjobs"],
    "lever": [r"lever\.co", r"jobs\.lever\.co"],
    "smartrecruiters": [r"smartrecruiters\.com", r"jobs\.smartrecruiters\.com"],
    "icims": [r"icims\.com", r"careers-.*\.icims\.com"],
    "taleo": [r"taleo\.net", r"oracle\.com/.*taleo"],
    "jobvite": [r"jobvite\.com", r"jobs\.jobvite\.com"],
    "bamboohr": [r"bamboohr\.com"],
    "linkedin": [r"linkedin\.com/jobs"],
    "indeed": [r"indeed\.com"],
    "glassdoor": [r"glassdoor\.com"],
}

_COMPILED_PATTERNS: Dict[str, list] = {
    board: [re.compile(p, re.I) for p in patterns]
    for board, patterns in JOB_BOARD_PATTERNS.items()
}

BOARD_FEATURES: Dict[str, Dict[str, bool]] = {
    "greenhouse": {
        "auto_fill_supported": True,
        "file_upload_supported": True,
        "custom_questions": True,
        "requires_manual_review": False,
    },
    "workday": {
        "auto_fill_supported": True,
        "file_upload_supported": True,
        "custom_questions": True,
        "requires_manual_review": True,
    },
    "lever": {
        "auto_fill_supported": True,
        "file_upload_supported": True,
        "custom_questions": True,
        "requires_manual_review": False,
    },
    "generic": {
        "auto_fill_supported": False,
        "file_upload_supported": False,
        "custom_questions": True,
        "requires_manual_review": True,
    },
}


def detect_job_board(url: str) -> str:
    """Detect the ATS platform from a URL. Returns board type or 'generic'."""
    if not url:
        return "generic"

    url_lower = url.lower()
    for board, patterns in _COMPILED_PATTERNS.items():
        for pattern in patterns:
            if pattern.search(url_lower):
                return board

    return "generic"


def get_job_board_info(url: str) -> Dict:
    """Get board type and feature support for a URL."""
    board_type = detect_job_board(url)
    features = BOARD_FEATURES.get(board_type, BOARD_FEATURES["generic"])

    return {
        "type": board_type,
        "features": features,
    }
